### PR TITLE
gloas: add new payload attestation processing

### PR DESCRIPTION
**Pull request by terencechain**
This PR implements [process_payload_attestation](https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/beacon-chain.md#new-process_payload_attestation) and spec tests

### CODE REVIEW

File: beacon-chain/core/gloas/payload_attestation.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+//	    selected.append(indices[next])
+//	  i += 1
+func selectByBalance(st state.ReadOnlyBeaconState, candidates []primitives.ValidatorIndex, seed [32]byte, count uint64) ([]primitives.ValidatorIndex, error) {
+	if len(candidates) == 0 {
+		return nil, errors.New("no candidates for balance weighted selection")
+	}
+
+	hashFunc := hash.CustomSHA256Hasher()
+	var buf [40]byte
+	copy(buf[:], seed[:])
```
Comment (terencechain) [inline beacon-chain/core/gloas/payload_attestation.go:159]: I did this so i can reuse it in `acceptByBalance` across  the for loop.  It's a little janky but it's performant
Will like feedback here

File: beacon-chain/core/gloas/payload_attestation.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+	return val.EffectiveBalance*fieldparams.MaxRandomValueElectra >= maxBalance*randomValue, nil
+}
+
+// validIndexedPayloadAttestation verifies the signature of an indexed payload attestation.
+// Spec v1.7.0-alpha.0 (pseudocode):
+// is_valid_indexed_payload_attestation(state: BeaconState, indexed_payload_attestation: IndexedPayloadAttestation) -> bool:
+//
+//	indices = indexed_payload_attestation.attesting_indices
+//	return len(indices) > 0 and indices == sorted(indices) and
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go:210]: more of a question on spec, i saw VerifyIndexedAttestation and it had

//	  if len(indices) == 0 or not indices == sorted(set(indices)):

does sorted indices not need a unqiueness check?

Comment (terencechain) [reply beacon-chain/core/gloas/payload_attestation.go:210]: No, for PTC, the way its sampled is different than beacon committee. It's sampled similarly to proposer, that you can represent multiple bits in the same message

Comment (james-prysm) [reply beacon-chain/core/gloas/payload_attestation.go:210]: thanks for that clarification

File: beacon-chain/core/gloas/payload_attestation.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+//	  bls.FastAggregateVerify(
+//	    [state.validators[i].pubkey for i in indices],
+//	    compute_signing_root(indexed_payload_attestation.data, get_domain(state, DOMAIN_PTC_ATTESTER, None)),
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go]: did this function get updated? i see in the specs it says
```

    # Verify aggregate signature
    pubkeys = [state.validators[i].pubkey for i in indices]
    domain = get_domain(state, DOMAIN_PTC_ATTESTER, compute_epoch_at_slot(attestation.data.slot))
    signing_root = compute_signing_root(attestation.data, domain)
    return bls.FastAggregateVerify(pubkeys, signing_root, attestation.signature)
    ```
    
https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/beacon-chain.md#new-is_valid_indexed_payload_attestation

File: beacon-chain/core/gloas/payload_attestation.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		keyBytes := val.PublicKey()
+		key, err := bls.PublicKeyFromBytes(keyBytes[:])
+		if err != nil {
+			return errors.Wrapf(err, "pubkey %d", idx)
+		}
+		pubkeys[i] = key
+	}
+
+	domain, err := signing.Domain(st.Fork(), slots.ToEpoch(st.Slot()), params.BeaconConfig().DomainPTCAttester, st.GenesisValidatorsRoot())
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go]: this should be using the attestation data slot 
```
domain = get_domain(state, DOMAIN_PTC_ATTESTER, compute_epoch_at_slot(attestation.data.slot))
```

File: beacon-chain/core/gloas/payload_attestation.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+//	offset = i % 16 * 2
+//	random_value = bytes_to_uint64(random_bytes[offset:offset+2])
+//	effective_balance = state.validators[index].effective_balance
+//	return effective_balance * MAX_RANDOM_VALUE >= MAX_EFFECTIVE_BALANCE_ELECTRA * random_value
+func acceptByBalance(st state.ReadOnlyBeaconState, idx primitives.ValidatorIndex, seedBuf []byte, hashFunc func([]byte) [32]byte, maxBalance uint64, round uint64) (bool, error) {
+	// Reuse the seed buffer by overwriting the last 8 bytes with the round counter.
+	binary.LittleEndian.PutUint64(seedBuf[len(seedBuf)-8:], round/16)
+	random := hashFunc(seedBuf)
+	offset := (round % 16) * 2
+	randomValue := uint64(binary.LittleEndian.Uint16(random[offset:])) // 16-bit draw per spec
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go]: would random_bytes[offset:offset+2] be more clear?

File: beacon-chain/core/gloas/payload_attestation.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	hashFunc := hash.CustomSHA256Hasher()
+	var buf [40]byte
+	copy(buf[:], seed[:])
+	maxBalance := params.BeaconConfig().MaxEffectiveBalanceElectra
+
+	selected := make([]primitives.ValidatorIndex, 0, count)
+	total := uint64(len(candidates))
+	for i := uint64(0); uint64(len(selected)) < count; i++ {
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go:164]: is it possible in tests for validators to all have effective balance 0 and cause an infinite loop?

File: beacon-chain/core/gloas/payload_attestation.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	seed, err := helpers.Seed(st, epoch, params.BeaconConfig().DomainPTCAttester)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return hash.Hash(append(seed[:], bytesutil.Bytes8(uint64(slot))...)), nil
+}
+
+// selectByBalance selects a balance-weighted subset of input candidates.
+// Spec v1.7.0-alpha.0 (pseudocode):
+// compute_balance_weighted_selection(state, indices, seed, size, shuffle_indices):
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go:141]: in our implementation we don't mention shuffle_indices false, maybe it's worth adding a comment below in the implementation

File: beacon-chain/core/gloas/payload_attestation.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		return nil, err
+	}
+
+	activeCount, err := helpers.ActiveValidatorCount(ctx, st, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	committeesPerSlot := helpers.SlotCommitteeCount(activeCount)
+	out := make([]primitives.ValidatorIndex, 0, committeesPerSlot*(activeCount/committeesPerSlot))
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go]: this capacity is wrong right? committeesPerSlot*(activeCount/committeesPerSlot) == activeCount...

File: beacon-chain/core/gloas/payload_attestation_test.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+		Signature:       signAttestation(t, st, attData, []common.SecretKey{sk2}),
+	}
+	body := buildBody(t, att)
+
+	err := ProcessPayloadAttestations(t.Context(), st, body)
+	require.ErrorContains(t, "failed to verify indexed form", err)
+	require.ErrorContains(t, "invalid signature", err)
+}
+
+func TestProcessPayloadAttestations_HappyPath(t *testing.T) {
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation_test.go:121]: what about multiple attestations?

File: beacon-chain/core/gloas/payload_attestation.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+//	  if compute_balance_weighted_acceptance(state, indices[next], seed, i):
+//	    selected.append(indices[next])
+//	  i += 1
+func selectByBalance(st state.ReadOnlyBeaconState, candidates []primitives.ValidatorIndex, seed [32]byte, count uint64) ([]primitives.ValidatorIndex, error) {
+	if len(candidates) == 0 {
+		return nil, errors.New("no candidates for balance weighted selection")
+	}
+
+	hashFunc := hash.CustomSHA256Hasher()
+	var buf [40]byte
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go:158]: might be worth adding a comment like
 // Pre-allocate buffer for hash input: seed (32 bytes) + round counter (8 bytes)

File: beacon-chain/core/gloas/payload_attestation.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/core/gloas/payload_attestation.go]: Yes this was updated yesterday, ill address it now

File: beacon-chain/core/gloas/payload_attestation.go (hunk 8 above)
Comment (terencechain) [reply beacon-chain/core/gloas/payload_attestation.go]: nice catch! addressed all your feedback in 3bcacd8a518c3ce15deed4512f846aec0ea7058e except for adding the infinity test

File: beacon-chain/core/gloas/payload_attestation_test.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+		Signature:       signAttestation(t, st, attData2, []common.SecretKey{sk2}),
+	}
+
+	body := buildBody(t, att1, att2)
+
+	err := ProcessPayloadAttestations(t.Context(), st, body)
+	require.NoError(t, err)
+}
+
+func TestProcessPayloadAttestations_IndexedVerificationError(t *testing.T) {
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation_test.go:190]: what about adding a test for empty aggregation bits, is that possible?

File: beacon-chain/core/gloas/payload_attestation.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+	hashFunc := hash.CustomSHA256Hasher()
+	// Pre-allocate buffer for hash input: seed (32 bytes) + round counter (8 bytes).
+	var buf [40]byte
+	copy(buf[:], seed[:])
+	maxBalance := params.BeaconConfig().MaxEffectiveBalanceElectra
+
+	selected := make([]primitives.ValidatorIndex, 0, count)
+	total := uint64(len(candidates))
+	for i := uint64(0); uint64(len(selected)) < count; i++ {
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation.go:164]: in NextSyncCommitteeIndices we have a check for ctx.Err(), why wouldn't these loops have that kind of check?

File: beacon-chain/core/gloas/payload_attestation_test.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+func newKey(t *testing.T) (common.SecretKey, []byte) {
+	sk, err := bls.RandKey()
+	require.NoError(t, err)
+	return sk, sk.PublicKey().Marshal()
+}
+
+func signAttestation(t *testing.T, st state.ReadOnlyBeaconState, data *eth.PayloadAttestationData, sks []common.SecretKey) []byte {
+	domain, err := signing.Domain(st.Fork(), slots.ToEpoch(st.Slot()), params.BeaconConfig().DomainPTCAttester, st.GenesisValidatorsRoot())
```
Comment (james-prysm) [inline beacon-chain/core/gloas/payload_attestation_test.go]: shouldn't this be data.slot ?

Comment (james-prysm) [reply beacon-chain/core/gloas/payload_attestation_test.go]: maybe the slots are the same so the tests pass but I think it should be based on the attestation data

**Review by james-prysm — DISMISSED** [review]
All the feedback I had, just need rebasing and another person's review/feedback

**Review by james-prysm — DISMISSED** [review]

File: beacon-chain/core/gloas/payload_attestation.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	header := st.LatestBlockHeader()
+
+	for i, att := range atts {
+		data := att.Data
+		if !bytes.Equal(data.BeaconBlockRoot, header.ParentRoot) {
+			return fmt.Errorf("payload attestation %d has wrong parent: got %x want %x", i, data.BeaconBlockRoot, header.ParentRoot)
+		}
+		if data.Slot+1 != st.Slot() {
```
Comment (potuz) [inline beacon-chain/core/gloas/payload_attestation.go]: `data.Slot` is untrusted here? couldn't the proposer include an attestation that will overflow here with the +1? At any rate, is there any danger with the overflow?

File: beacon-chain/core/gloas/payload_attestation.go [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+		return nil, err
+	}
+
+	activeCount, err := helpers.ActiveValidatorCount(ctx, st, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	committeesPerSlot := helpers.SlotCommitteeCount(activeCount)
+	out := make([]primitives.ValidatorIndex, 0, committeesPerSlot)
```
Comment (potuz) [inline beacon-chain/core/gloas/payload_attestation.go]: This is almost clear to be reallocating: you are allocating `committeesPerSlot` validator indices in this slice, but then copying a full committee into it.

File: beacon-chain/core/gloas/payload_attestation.go (hunk 14 above)
Comment (terencechain) [reply beacon-chain/core/gloas/payload_attestation.go]: I think you are right that p2p spec does not guard against this, a malicious proposer could trigger the overflow here, and the worst case is consensus spec that we treat payload attestation as valid but other rejects it. I'll add the overflow check to be safe

File: beacon-chain/core/gloas/payload_attestation.go (hunk 15 above)
Comment (terencechain) [reply beacon-chain/core/gloas/payload_attestation.go]: Very good catch

File: beacon-chain/core/gloas/payload_attestation.go [hunk 16]
```diff
@@ -48,7 +48,7 @@ func ProcessPayloadAttestations(ctx context.Context, st state.BeaconState, body
 		if !bytes.Equal(data.BeaconBlockRoot, header.ParentRoot) {
 			return fmt.Errorf("payload attestation %d has wrong parent: got %x want %x", i, data.BeaconBlockRoot, header.ParentRoot)
 		}
-		if data.Slot+1 != st.Slot() {
+		if data.Slot.Add(1) != st.Slot() {
```
Comment (potuz) [inline beacon-chain/core/gloas/payload_attestation.go]: This is strictly worse this panics!

**Review by potuz — CHANGES_REQUESTED** [review]

**Review by james-prysm — APPROVED** [review]
approving since I think all my concerns and my interpretation of potuz's concerns have been addressed

**Review by potuz — APPROVED** [review]
